### PR TITLE
feat: notification for unread chats scrolled out of view

### PR DESCRIPTION
**Pull request by suprko**
#### Relevant info and issue/PR links 
https://app.clubhouse.io/peerio/story/838/desktop-unread-message-out-of-view-awareness
one text entry added to copy: https://github.com/PeerioTechnologies/peerio-icebear/pull/150

#### Testing instructions
Intended behav:
1. When there are unread messages scrolled out of view in ChatList, a little UI thingie shows up and points you towards unread messages, either at the top or bottom of ChatList depending on where the unreads are.
2. When there are unread messages above *and* below, this notification only shows up at the top. This is designers' decision, don't @ me.
3. Clicking on this notification causes ChatList to scroll until the next unread message is onscreen.

----
### Repository owner

- [ ] Was tested and can be merged.
- [ ] PR name follows conventional changelog format.

[PR guideline](https://github.com/PeerioTechnologies/peerio-icebear/blob/dev/docs/CONTRIBUTING.md)

### CODE REVIEW

**Comment by suprko** [comment]
@anri-asaturov when you get a chance, I've got a couple of questions/concerns about this (see PR comment)

**Comment by anri-asaturov** [comment]
Ok. I'm gonna review this now, but generally I'd prefer discussing this kind of things before implementation.

File: src/ui/chat/components/ChatList.jsx [hunk 1]
```diff
@@ -1,27 +1,287 @@
...
+    }
+
+    componentWillUnmount() {
+        window.removeEventListener('resize', this.checkScrollHeight);
+        this.scrollContainerRef.removeEventListener('scroll', this.checkUnreadsOffscreen);
+        window.removeEventListener('resize', this.checkUnreadsOffscreen);
+    }
+
+    // Building the rooms & invites list
+    @computed get allRooms() {
```
Comment (suprko) [inline src/ui/chat/components/ChatList.jsx]: Construction of the ChatList (rooms, room invites, and DMs) has been moved out of `render` to `computed` functions, which (I think?) is more performant but more importantly also allows them to be accessed by the other functions

File: src/ui/chat/components/ChatList.jsx [hunk 2]
```diff
@@ -40,47 +276,7 @@ class ChatList extends React.Component {
         routerStore.navigateTo(routerStore.ROUTES.newChannel);
     }
 
-    inviteOptions(kegDbId) {
```
Comment (suprko) [inline src/ui/chat/components/ChatList.jsx]: All of these functions were removed, some were just leftover unused functions from previous ChannelInvites design, others are redundant one-liners that are better as in-line JSX.

File: src/ui/chat/components/ChatList.jsx [hunk 3]
```diff
@@ -22,6 +31,233 @@ class ChatList extends React.Component {
...
+        if (this.prevUnread) {
+            this.goToUnread = this.prevUnread;
+            window.requestAnimationFrame(this.smoothScrollUp);
+        } else {
+            this.goToUnread = this.nextUnread;
+            window.requestAnimationFrame(this.smoothScrollDown);
+        }
+    }
+
+    smoothScrollUp = () => {
```
Comment (suprko) [inline src/ui/chat/components/ChatList.jsx]: "smoothScroll" functions borrowed from the one used in message list

**Comment by suprko** [comment]
@anri-asaturov ready for review

1. Reworked so that it doesn't calculate any DOM `rect`s at all, just calculates ChatList `<li>` positions based on hardcoded padding & height values
2. Scroll-to-next/prev-unread on click works now
3. The unrelated edits (where I removed a bunch of unused functions) are still there. I marked that section in the comments.

File: src/ui/chat/components/ChatList.jsx [hunk 4]
```diff
@@ -22,6 +31,233 @@ class ChatList extends React.Component {
         }
     }
 
+    componentWillUnmount() {
+        window.removeEventListener('resize', this.checkScrollHeight);
+        this.scrollContainerRef.removeEventListener('scroll', this.checkScrollDistance);
```
Comment (anri-asaturov) [inline src/ui/chat/components/ChatList.jsx]: Check for null ref just in case.

File: src/ui/chat/components/ChatList.jsx [hunk 5]
```diff
@@ -22,6 +31,233 @@ class ChatList extends React.Component {
...
+        if (!ref) {
+            this.scrollContainerRef = null;
+            return;
+        }
+
+        this.scrollContainerRef = ref;
+        this.checkScrollHeight();
+        window.addEventListener('resize', this.checkScrollHeight);
+
+        ref.addEventListener('scroll', this.checkScrollDistance);
```
Comment (anri-asaturov) [inline src/ui/chat/components/ChatList.jsx]: Throttle this. There will be too many scroll events and each of them starts the whole recalculation.

File: src/ui/chat/components/ChatList.jsx [hunk 6]
```diff
@@ -22,6 +31,233 @@ class ChatList extends React.Component {
...
+            );
+        });
+    }
+
+    // Calculating the positions of unread messages relative to scroll container
+    @computed get unreadPositions() {
+        const positionsArray = [];
+
+        const unreadRooms = this.allRooms.reduce((acc, room, i) => {
+            if (room.unreadCount > 0) {
```
Comment (anri-asaturov) [inline src/ui/chat/components/ChatList.jsx]: I think we should skip active chats, or the button will briefly show for them with every new message,

File: src/ui/chat/components/ChatList.jsx [hunk 7]
```diff
@@ -22,6 +31,233 @@ class ChatList extends React.Component {
...
+
+        const roomsOffset = paddingTop + buttonHeight;
+        unreadRooms.forEach(i => {
+            positionsArray.push(
+                roomsOffset + i * roomHeight
+            );
+        });
+
+        const unreadDMs = chatStore.directMessages.reduce((acc, chat, i) => {
+            if (chat.unreadCount > 0) {
```
Comment (anri-asaturov) [inline src/ui/chat/components/ChatList.jsx]: Here too - skip active chat

**Comment by suprko** [comment]
@anri-asaturov all fixed

**Comment by ebilg** [comment]
can be merged

**Pull request merged**